shakenfist/scheduler.py

```diff
@@ -179,11 +194,17 @@ def _find_most_matching_images(self, requested_images, candidates):
 
     def place_instance(self, instance, network, candidates=None):
         with util.RecordedOperation('schedule', instance):
-            log_ctx = LOG.with_object(instance)
+            log_ctx = self.log.with_object(instance)
 
+            # Refresh metrics if its too old, or there are no nodes.
             diff = time.time() - self.metrics_updated
-            if diff > config.get('SCHEDULER_CACHE_TIMEOUT'):
+            log_ctx.debug('Metrics are %.02f seconds old, max is %.02f seconds. '
+                          'Cache has %d elements.'
+                          % (diff, config.SCHEDULER_CACHE_TIMEOUT, len(self.metrics)))
+            if diff > config.SCHEDULER_CACHE_TIMEOUT or len(self.metrics) == 0:
                 self.refresh_metrics()
+                log_ctx.debug('Cache has %d elements after refresh.'
+                              % len(self.metrics))
```

Comment:
For efficiency, log messages should use the args and kwargs of debug() rather sending an evaluated string. The logger delays evaluation until it checks the logging level.

Reply:
Done.